zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2234,13 +2205,59 @@ private String getHexSessionId() {
   }
 
   /*
-   * Session aware operation needs below requirements:
-   * 1. the session id is NOT null or empty
-   * 2. create mode is EPHEMERAL or EPHEMERAL_SEQUENTIAL
+   * Gets the zookeeper instance that ensures its session ID matches the expected session ID.
+   * It is used for write operations that suppose the znode to be created by the expected session.
    */
-  private boolean isSessionAwareOperation(String expectedSessionId, CreateMode mode) {
-    return expectedSessionId != null && !expectedSessionId.isEmpty() && (
-        mode == CreateMode.EPHEMERAL || mode == CreateMode.EPHEMERAL_SEQUENTIAL);
+  private ZooKeeper getExpectedZookeeper(final String expectedSessionId) {
+    /*
+     * Cache the zookeeper reference and make sure later zooKeeper.create() is being run
+     * under this zookeeper connection. This is to avoid zk session change after expected
+     * session check.
+     */
+    ZooKeeper zk = ((ZkConnection) getConnection()).getZookeeper();
+
+    /*
+     * The operation is NOT session aware, we will use the actual zookeeper session without
+     * checking expected session.
+     */
+    if (expectedSessionId == null || expectedSessionId.isEmpty()) {
+      return zk;
+    }
+
+    /*
+     * If operation is session aware (expectedSession is valid),
+     * we have to check whether or not the passed-in(expected) session id
+     * matches actual session's id.
+     * If not, we should not return a zk object for the zk operation.
+     */
+    final String actualSessionId = Long.toHexString(zk.getSessionId());
+    if (!actualSessionId.equals(expectedSessionId)) {
+      throw new ZkSessionMismatchedException(
+          "Failed to get expected zookeeper instance! There is a session id mismatch. Expected: "
+              + expectedSessionId + ". Actual: " + actualSessionId);
+    }
+
+    return zk;
+  }
+
+  private String parseExpectedSessionId(Object data) {
+    if (data == null) {
+      return null;
+    }
+
+    ZNRecord record;
+    try {
+      record = (ZNRecord) data;
```

Comment:
I've thought about it. `data instanceof Message` adds one more check.
```
if (!data instanceof Message) {
  return null;
}

ZNRecord record = (ZNRecord) data;
...
```

Since most of the time, this API is for message and we need to cast the type, I prefer this so no need to always do the `instanceof` check.